File: src/scoring.cpp
// scoring.cpp - VEX V5RC Override scoring node.
//   - Each placed pin has two terminals (top / bottom color halves).
//   - A terminal is worth 5 pts if red/blue, 10 pts if yellow.
//   - A terminal is hidden if the cup stacked on that pin has its opaque half
//     facing it (simplified: cup yaw in [0, pi) hides the bottom terminal,
//     [pi, 2pi) hides the top terminal).
//   - Yellow pins in a quadrant are owned by the alliance whose color matches
//     the quadrant Toggle; yellow pins on the center (midfield) goal are owned
//     by the alliance with more robots in the Midfield (tie = no one).
//   - A robot with any part inside the Midfield scores 8 pts.
//   - Autonomous bonus: +12 to the alliance with more autonomous points
//     (pins only), +6 each on a tie.
#include <algorithm>
#include <cmath>
#include <cstdint>
#include <memory>
#include <utility>
#include <vector>

#include <geometry_msgs/msg/pose_array.hpp>
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/int64_multi_array.hpp>

#include "override_sim/msg/cup_element.hpp"
#include "override_sim/msg/goal_array.hpp"
#include "override_sim/msg/goal_state.hpp"
#include "override_sim/msg/pin_element.hpp"

namespace override_sim
{

constexpr int kPinTerminalPoints = 5;
constexpr int kYellowPinPoints = 10;
constexpr int kRobotInMidfieldPoints = 8;
constexpr int kAutonBonus = 12;
constexpr int kAutonBonusTie = 6;

constexpr double kMidfieldHalf = 0.6;   // midfield diamond |x|+|y| <= 0.6 m
constexpr double kRobotRadius = 0.15;   // any part of the robot counts

// one entry of a goal's element stack, sorted bottom-up
struct StackEntry
{
  double z;
  int kind;          // 0 = pin, 1 = cup
  PinElement pin;
  CupElement cup;
};

class Scoring : public rclcpp::Node
{
public:
  Scoring() : Node("scoring")
  {
    goals_sub_ = this->create_subscription<GoalArray>(
      "/goals", 10,
      [this](const GoalArray::SharedPtr msg) { score_calculator(msg); });
    robot_pose_sub_ = this->create_subscription<geometry_msgs::msg::PoseArray>(
      "/otto_pose", 10,
      [this](const geometry_msgs::msg::PoseArray::SharedPtr msg) {
        if (!msg->poses.empty()) {
          robot_x_ = msg->poses.back().position.x;
          robot_y_ = msg->poses.back().position.y;
        }
      });
    opponent_pose_sub_ =
      this->create_subscription<geometry_msgs::msg::PoseStamped>(
      "/opponent/pose", 10,
      [this](const geometry_msgs::msg::PoseStamped::SharedPtr msg) {
        opp_x_ = msg->pose.position.x;
        opp_y_ = msg->pose.position.y;
      });
    phase_sub_ = this->create_subscription<std_msgs::msg::Int64MultiArray>(
      "/game_phase", 10,
      [this](const std_msgs::msg::Int64MultiArray::SharedPtr msg) {
        phase_cb(msg);
      });

    score_pub_ = this->create_publisher<std_msgs::msg::Int64MultiArray>(
      "/game_score", 10);
    detail_pub_ = this->create_publisher<std_msgs::msg::Int64MultiArray>(
      "/score_detail", 10);
  }

private:
  void phase_cb(const std_msgs::msg::Int64MultiArray::SharedPtr msg)
  {
    if (msg->data.empty()) {
      return;
    }
    const int64_t phase = msg->data[0];
    if (phase == 0) {
      if (!auton_started_) {
        auton_started_ = true;
        auton_active_ = true;
        auton_red_ = 0;
        auton_blue_ = 0;
        auton_bonus_red_ = 0;
        auton_bonus_blue_ = 0;
        RCLCPP_INFO(this->get_logger(), "autonomous period started");
      }
    } else if (phase == 1 && auton_active_) {
      auton_active_ = false;
      if (auton_red_ > auton_blue_) {
        auton_bonus_red_ = kAutonBonus;
        auton_bonus_blue_ = 0;
      } else if (auton_blue_ > auton_red_) {
        auton_bonus_red_ = 0;
        auton_bonus_blue_ = kAutonBonus;
      } else {
        auton_bonus_red_ = kAutonBonusTie;
        auton_bonus_blue_ = kAutonBonusTie;
      }
      RCLCPP_INFO(this->get_logger(),
        "autonomous bonus: red %d, blue %d", auton_bonus_red_,
        auton_bonus_blue_);
    }
  }

  static bool InMidfield(double x, double y)
  {
    return std::abs(x) + std::abs(y) <= kMidfieldHalf + kRobotRadius;
  }

  std::pair<int, int> midfield_counts() const
  {
    const int red = InMidfield(robot_x_, robot_y_) ? 1 : 0;
    const int blue = InMidfield(opp_x_, opp_y_) ? 1 : 0;
    return {red, blue};
  }

  // 1 (red), 2 (blue) or 0 (no one) for a yellow terminal on this goal
  int yellow_owner(const GoalState & gs, int red_mid, int blue_mid) const
  {
    if (gs.quadrant == "C") {
      if (red_mid > blue_mid) return 1;
      if (blue_mid > red_mid) return 2;
      return 0;
    }
    return gs.toggle_state;
  }

  // 0 = none, 1 = bottom terminal hidden, 2 = top terminal hidden
  static int cup_covering_half(const PinElement & pin,
    const std::vector<StackEntry> & stack)
  {
    double best_d = 0.0;
    bool found = false;
    const CupElement * covering = nullptr;
    for (const auto & e : stack) {
      if (e.kind != 1) {
        continue;
      }
      const double dz = e.z - pin.location.z;
      if (dz >= 0.10 && dz <= 0.28) {
        if (!found || dz < best_d) {
          found = true;
          best_d = dz;
          covering = &e.cup;
        }
      }
    }
    if (!found || covering == nullptr) {
      return 0;
    }
    return covering->yaw < M_PI ? 1 : 2;
  }

  void score_calculator(const GoalArray::SharedPtr goals)
  {
    int red = 0;
    int blue = 0;
    int red_pins = 0;
    int blue_pins = 0;
    int red_yellow = 0;
    int blue_yellow = 0;

    const auto mids = midfield_counts();
    const int red_mid = mids.first;
    const int blue_mid = mids.second;

    for (const auto & gs : goals->goals) {
      // combine pins and cups into one stack, sorted bottom-up
      std::vector<StackEntry> stack;
      for (const auto & pin : gs.pins) {
        StackEntry e;
        e.z = pin.location.z;
        e.kind = 0;
        e.pin = pin;
        stack.push_back(e);
      }
      for (const auto & cup : gs.cups) {
        StackEntry e;
        e.z = cup.location.z;
        e.kind = 1;
        e.cup = cup;
        stack.push_back(e);
      }
      std::sort(stack.begin(), stack.end(),
        [](const StackEntry & a, const StackEntry & b) { return a.z < b.z; });

      for (const auto & e : stack) {
        if (e.kind != 0) {
          continue;
        }
        const PinElement & pin = e.pin;
        const int covered = cup_covering_half(pin, stack);
        const std::vector<int8_t> terminals = {
          covered == 2 ? 0 : pin.top_color,
          covered == 1 ? 0 : pin.bottom_color,
        };

        for (const int8_t color : terminals) {
          if (color == 1) {
            red += kPinTerminalPoints;
            red_pins += kPinTerminalPoints;
          } else if (color == 2) {
            blue += kPinTerminalPoints;
            blue_pins += kPinTerminalPoints;
          } else if (color == 3) {
            const int owner = yellow_owner(gs, red_mid, blue_mid);
            if (owner == 1) {
              red += kYellowPinPoints;
              red_yellow += kYellowPinPoints;
            } else if (owner == 2) {
              blue += kYellowPinPoints;
              blue_yellow += kYellowPinPoints;
            }
          }
        }
      }
    }

    // robots in the midfield
    red += red_mid * kRobotInMidfieldPoints;
    blue += blue_mid * kRobotInMidfieldPoints;

    // autonomous pin-score snapshot (midfield points excluded per SC7)
    if (auton_active_) {
      auton_red_ = red_pins + red_yellow;
      auton_blue_ = blue_pins + blue_yellow;
    }

    red += auton_bonus_red_;
    blue += auton_bonus_blue_;

    auto score_msg = std::make_unique<std_msgs::msg::Int64MultiArray>();
    score_msg->data.push_back(red);
    score_msg->data.push_back(blue);
    score_pub_->publish(std::move(score_msg));

    auto detail = std::make_unique<std_msgs::msg::Int64MultiArray>();
    detail->data.push_back(red_pins);
    detail->data.push_back(blue_pins);
    detail->data.push_back(red_yellow);
    detail->data.push_back(blue_yellow);
    detail->data.push_back(red_mid * kRobotInMidfieldPoints);
    detail->data.push_back(blue_mid * kRobotInMidfieldPoints);
    detail->data.push_back(auton_bonus_red_);
    detail->data.push_back(auton_bonus_blue_);
    detail_pub_->publish(std::move(detail));
  }

  // state
  bool auton_active_ = false;
  bool auton_started_ = false;
  int auton_bonus_red_ = 0;
  int auton_bonus_blue_ = 0;
  int auton_red_ = 0;
  int auton_blue_ = 0;
  double robot_x_ = 0.0;
  double robot_y_ = 0.0;
  double opp_x_ = 0.0;
  double opp_y_ = 0.0;

  rclcpp::Subscription<GoalArray>::SharedPtr goals_sub_;
  rclcpp::Subscription<geometry_msgs::msg::PoseArray>::SharedPtr robot_pose_sub_;
  rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr
    opponent_pose_sub_;
  rclcpp::Subscription<std_msgs::msg::Int64MultiArray>::SharedPtr phase_sub_;
  rclcpp::Publisher<std_msgs::msg::Int64MultiArray>::SharedPtr score_pub_;
  rclcpp::Publisher<std_msgs::msg::Int64MultiArray>::SharedPtr detail_pub_;
};

}  // namespace override_sim

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<override_sim::Scoring>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

File: src/scoring.py
#!/usr/bin/env python3

import math

import rclpy
from rclpy.node import Node
from std_msgs.msg import Int64MultiArray

from geometry_msgs.msg import PoseArray, PoseStamped

from override_sim.msg import GoalArray

# color codes: 1 = red, 2 = blue, 3 = yellow

PIN_TERMINAL_POINTS = 5
YELLOW_PIN_POINTS = 10
ROBOT_IN_MIDFIELD_POINTS = 8
AUTON_BONUS = 12
AUTON_BONUS_TIE = 6

MIDFIELD_HALF = 0.6       # midfield diamond |x|+|y| <= 0.6 m
ROBOT_RADIUS = 0.15       # any part of the robot counts -> inflate by this


class Scoring(Node):
    """VEX V5RC Override scoring.

    - Each placed pin has two terminals (top / bottom color halves).
    - A terminal is worth 5 pts if red/blue, 10 pts if yellow.
    - A terminal is hidden if the cup stacked on that pin has its opaque half
      facing it (simplified: cup yaw in [0, pi) hides the bottom terminal,
      [pi, 2pi) hides the top terminal).
    - Yellow pins in a quadrant are owned by the alliance whose color matches
      the quadrant Toggle; yellow pins on the center (midfield) goal are owned
      by the alliance with more robots in the Midfield (tie = no one).
    - A robot with any part inside the Midfield scores 8 pts.
    - Autonomous bonus: +12 to the alliance with more autonomous points
      (pins only), +6 each on a tie.
    """

    def __init__(self):
        super().__init__('scoring')

        self.create_subscription(GoalArray, '/goals', self.score_calculator, 10)
        self.create_subscription(PoseArray, '/otto_pose', self.robot_pose_cb, 10)
        self.create_subscription(
            PoseStamped, '/opponent/pose', self.opponent_pose_cb, 10)
        self.create_subscription(
            Int64MultiArray, '/game_phase', self.phase_cb, 10)

        self.score = self.create_publisher(Int64MultiArray, '/game_score', 10)
        self.score_detail = self.create_publisher(
            Int64MultiArray, '/score_detail', 10)

        # match phase state
        self.auton_active = False
        self.auton_started = False
        self.auton_bonus = (0, 0)
        self.auton_red = 0
        self.auton_blue = 0

        self.robot_x, self.robot_y = 0.0, 0.0
        self.opp_x, self.opp_y = 0.0, 0.0

    # ------------------------------------------------------------------
    def phase_cb(self, msg):
        phase = int(msg.data)
        if phase == 0:
            if not self.auton_started:
                self.auton_started = True
                self.auton_active = True
                self.auton_red = 0
                self.auton_blue = 0
                self.auton_bonus = (0, 0)
                self.get_logger().info("autonomous period started")
        elif phase == 1 and self.auton_active:
            self.auton_active = False
            if self.auton_red > self.auton_blue:
                self.auton_bonus = (AUTON_BONUS, 0)
            elif self.auton_blue > self.auton_red:
                self.auton_bonus = (0, AUTON_BONUS)
            else:
                self.auton_bonus = (AUTON_BONUS_TIE, AUTON_BONUS_TIE)
            self.get_logger().info(
                f"autonomous bonus: red {self.auton_bonus[0]}, "
                f"blue {self.auton_bonus[1]}")

    def robot_pose_cb(self, msg: PoseArray):
        self.robot_x = msg.poses[-1].position.x
        self.robot_y = msg.poses[-1].position.y

    def opponent_pose_cb(self, msg: PoseStamped):
        self.opp_x = msg.pose.position.x
        self.opp_y = msg.pose.position.y

    # ------------------------------------------------------------------
    @staticmethod
    def in_midfield(x, y):
        return abs(x) + abs(y) <= MIDFIELD_HALF + ROBOT_RADIUS

    def midfield_counts(self):
        red = 1 if self.in_midfield(self.robot_x, self.robot_y) else 0
        blue = 1 if self.in_midfield(self.opp_x, self.opp_y) else 0
        return red, blue

    def yellow_owner(self, gs, red_mid, blue_mid):
        """Return 1 (red), 2 (blue) or 0 (no one) for a yellow terminal."""
        if gs.quadrant == 'C':        # midfield goal
            if red_mid > blue_mid:
                return 1
            if blue_mid > red_mid:
                return 2
            return 0
        return gs.toggle_state

    def score_calculator(self, goals: GoalArray):
        red = 0
        blue = 0
        red_pins = 0
        blue_pins = 0
        red_yellow = 0
        blue_yellow = 0

        red_mid, blue_mid = self.midfield_counts()

        for gs in goals.goals:
            # combine pins and cups into one stack, sorted bottom-up
            stack = []
            for pin in gs.pins:
                stack.append((pin.location.z, 'pin', pin))
            for cup in gs.cups:
                stack.append((cup.location.z, 'cup', cup))
            stack.sort(key=lambda item: item[0])

            for base_z, kind, obj in stack:
                if kind != 'pin':
                    continue
                covered_half = self.cup_covering_half(obj, stack)

                terminals = []
                if covered_half != 'top':
                    terminals.append(obj.top_color)
                if covered_half != 'bottom':
                    terminals.append(obj.bottom_color)

                for color in terminals:
                    if color == 1:
                        red += PIN_TERMINAL_POINTS
                        red_pins += PIN_TERMINAL_POINTS
                    elif color == 2:
                        blue += PIN_TERMINAL_POINTS
                        blue_pins += PIN_TERMINAL_POINTS
                    elif color == 3:
                        owner = self.yellow_owner(gs, red_mid, blue_mid)
                        if owner == 1:
                            red += YELLOW_PIN_POINTS
                            red_yellow += YELLOW_PIN_POINTS
                        elif owner == 2:
                            blue += YELLOW_PIN_POINTS
                            blue_yellow += YELLOW_PIN_POINTS

        # robots in the midfield
        red += red_mid * ROBOT_IN_MIDFIELD_POINTS
        blue += blue_mid * ROBOT_IN_MIDFIELD_POINTS

        # autonomous pin-score snapshot (midfield points excluded per SC7)
        if self.auton_active:
            self.auton_red = red_pins + red_yellow
            self.auton_blue = blue_pins + blue_yellow

        red += self.auton_bonus[0]
        blue += self.auton_bonus[1]

        score_array = Int64MultiArray()
        score_array.data = [red, blue]
        self.score.publish(score_array)

        detail = Int64MultiArray()
        detail.data = [red_pins, blue_pins, red_yellow, blue_yellow,
                       red_mid * ROBOT_IN_MIDFIELD_POINTS,
                       blue_mid * ROBOT_IN_MIDFIELD_POINTS,
                       self.auton_bonus[0], self.auton_bonus[1]]
        self.score_detail.publish(detail)

    def cup_covering_half(self, pin, stack):
        """Return 'top'/'bottom'/'none': which pin terminal the nearest cup
        above this pin hides. Cup yaw in [0, pi) hides the bottom terminal;
        [pi, 2pi) hides the top terminal (simplified SC3 model)."""
        best_d = None
        covering_cup = None
        for cup_z, kind, obj in stack:
            if kind != 'cup':
                continue
            dz = cup_z - pin.location.z
            if 0.10 <= dz <= 0.28:
                if best_d is None or dz < best_d:
                    best_d = dz
                    covering_cup = obj
        if covering_cup is None:
            return None
        return 'bottom' if covering_cup.yaw < math.pi else 'top'


def main(args=None):
    rclpy.init(args=args)
    node = Scoring()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == '__main__':
    main()
